File: diploma-server/server-app/src/test/java/org/arhor/diploma/data/persistence/repository/AccountRepositoryTest.java
package org.arhor.diploma.data.persistence.repository;

import static org.assertj.core.api.Assertions.assertThat;

import java.util.Optional;

import org.arhor.diploma.data.persistence.domain.Account;
import org.arhor.diploma.testutils.RandomParameter;
import org.junit.jupiter.api.Test;
import org.junit.platform.commons.annotation.Testable;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.transaction.annotation.Transactional;

@Testable
class AccountRepositoryTest extends DatabaseIntegrationTest {

    @Autowired
    private AccountRepository repository;

    @Test
    @Transactional
    void shouldSoftlyDeleteAccount(@RandomParameter final Account account) {
        // given
        account.setId(null);

        final var savedAccount = repository.save(account);
        final var accountId = savedAccount.getId();

        // when
        repository.delete(savedAccount);

        // then
        assertThat(accountId).isNotNull();
        assertThat(repository.findById(accountId)).isEmpty();
        assertThat(repository.findDeletedIds()).contains(accountId);
    }

    @Test
    @Transactional
    void shouldSoftlyDeleteAccountById(@RandomParameter final Account account) {
        // given
        account.setId(null);

        final var savedAccount = repository.save(account);
        final var accountId = savedAccount.getId();

        // when
        repository.deleteById(accountId);

        // then
        assertThat(accountId).isNotNull();
        assertThat(repository.findById(accountId)).isEmpty();
        assertThat(repository.findDeletedIds()).contains(accountId);
    }

    @Test
    @Transactional
    void shouldFindAccountByUsername(@RandomParameter final Account account) {
        // given
        account.setId(null);

        final var savedAccount = repository.save(account);
        final var savedAccountUsername = savedAccount.getUsername();

        // when
        Optional<Account> optionalAccount = repository.findByUsername(savedAccountUsername);

        // then
        assertThat(optionalAccount).contains(savedAccount);
    }
}

File: diploma-server/server-app/src/test/kotlin/org/arhor/diploma/data/persistence/repository/AccountRepositoryTest.kt
package org.arhor.diploma.data.persistence.repository

import kotlinx.coroutines.flow.toList
import kotlinx.coroutines.runBlocking
import org.arhor.diploma.data.persistence.domain.Account
import org.arhor.diploma.testutils.RandomParameter
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired

class AccountRepositoryTest : DatabaseIntegrationTest() {

    @Autowired
    private lateinit var repository: AccountRepository

    @Test
    fun `should softly delete account`(@RandomParameter account: Account): Unit = runBlocking {
        // given
        account.id = null

        val savedAccount = repository.save(account)
        val accountId = savedAccount.id

        // when
        repository.delete(savedAccount)
        val resultById = repository.findById(accountId!!)
        val deletedAccountsIds = repository.findDeletedIds().toList()

        // then
        assertThat(accountId).isNotNull
        assertThat(resultById).isNull()
        assertThat(deletedAccountsIds).contains(accountId)
    }

    @Test
    fun `should softly delete account by id`(@RandomParameter account: Account): Unit = runBlocking {
        // given
        account.id = null

        val savedAccount = repository.save(account)
        val accountId = savedAccount.id

        // when
        repository.deleteById(accountId!!)
        val resultById = repository.findById(accountId)
        val deletedAccountsIds = repository.findDeletedIds().toList()

        // then
        assertThat(accountId).isNotNull
        assertThat(resultById).isNull()
        assertThat(deletedAccountsIds).contains(accountId)
    }

    @Test
    fun `should find account by username`(@RandomParameter account: Account): Unit = runBlocking {
        // given
        val savedAccount = repository.save(account.apply { id = null })

        // when
        val optionalAccount = repository.findByUsername(savedAccount.username!!)

        // then
        assertThat(optionalAccount)
            .isNotNull
            .isEqualTo(savedAccount)
    }
}
